coding/multilang_utf8_string.cpp

```diff
@@ -209,48 +209,30 @@ bool StringUtf8Multilang::HasString(int8_t lang) const
   return false;
 }
 
-namespace
+int8_t StringUtf8Multilang::FindString(string const & utf8s) const
 {
+  int8_t result = kUnsupportedLanguageCode;
 
-struct Printer
-{
-  string & m_out;
-  Printer(string & out) : m_out(out) {}
-  bool operator()(int8_t code, string const & name) const
-  {
-    m_out += string(StringUtf8Multilang::GetLangByCode(code)) + string(":") + name + " ";
-    return true;
-  }
-};
-
-struct Finder
-{
-  string const & m_s;
-  int8_t m_res;
-  Finder(string const & s) : m_s(s), m_res(-1) {}
-  bool operator()(int8_t code, string const & name)
-  {
-    if (name == m_s)
+  ForEach([&utf8s, &result](int8_t code, string const & name) -> base::ControlFlow {
+    if (name == utf8s)
     {
-      m_res = code;
-      return false;
+      result = code;
+      return base::ControlFlow::Break;
     }
-    return true;
-  }
-};
+    return base::ControlFlow::Continue;
+  });
 
-} // namespace
-
-int8_t StringUtf8Multilang::FindString(string const & utf8s) const
-{
-  Finder finder(utf8s);
-  ForEach(finder);
-  return finder.m_res;
+  return result;
 }
 
 string DebugPrint(StringUtf8Multilang const & s)
 {
-  string out;
-  s.ForEach(Printer(out));
-  return out;
+  string result;
+
+  s.ForEach([&result](int8_t code, string const & name) -> base::ControlFlow {
```

Comment:
I suggest you to implement two versions of ForEach in StringUtf8Multilang - the first one accepts lambdas that return void and the second one accepts lambdas that returns ControlFlow. Other way is:
```c++
template <typaname Fn>
void ForEach(Fn && fn)
{
  ControlFlowWrapper<Fn> wrapper(forward<Fn>(fn));
  for (...)
  {
     if (wrapper(args) == Break)
       break;
  }
}
```
I.e. ControlFlowWrapper forwards args to fn and returns Continue for functions that returns void and returns result value of functions that return ControlFlow.

Reply:
done